### ISSUE TITLE
how to disable the assembly optimization

### ISSUE THREAD

**Issue by XiWeiGu** [xianyi/OpenBLAS#2368]
If make the source code directly, it will automatically detecting architecture and open the corresponding assembly optimization. what should i do to use the pure c/c++ code instead of the assembly optimization?

**Comment by martin-frbg** [comment]
make `TARGET=GENERIC`to get a pure C build, or if you want to disable a particular assembly function (e.g. SAXPY) edit the KERNEL._TARGET_ file for your autodetected target (e.g. KERNEL.LOONGSON3A in kernel/mips64) and replace the assembly file listed there (the SAXPYKERNEL=) with the  corresponding C source from kernel/generic or kernel/arm

**Comment by XiWeiGu** [comment]
Use `make TARGET=GENERIC` in loongson 3A4000 (mips64), i got the following error message:
`make[1]: *** No rule to make target 'dgemm_incopy.o', needed by 'libs'。 停止.`   
`make[1]: Leaving directory '/data/gxw/community/OpenBLAS/kernel' `
`Makefile:142: recipe for target 'libs' failed`
`make: *** [libs] Error 1`
I found the `SGEMM_UNROLL_M` is equal to `SGEMM_UNROLL_N` in Makefile.conf, so there is no rule to make target dgemm_incopy.o

**Comment by martin-frbg** [comment]
For a generic target, DGEMM_UNROLL_M and DGEMM_UNROLL_N are both set to 2 (in param.h), as the only provided C kernel is generic/gemmkernel_2x2.c. This indeed does not require INCOPY/ITCOPY kernels, just
```
DGEMMKERNEL    =  ../generic/gemmkernel_2x2.c
DGEMMONCOPY    = ../generic/gemm_ncopy_2.c
DGEMMOTCOPY    = ../generic/gemm_tcopy_2.c
DGEMMONCOPYOBJ = dgemm_oncopy.o
DGEMMOTCOPYOBJ = dgemm_otcopy.o
```
(as in kernel/x86/KERNEL.generic), perhaps the build system is broken for a truly generic mips64 target, I suspect MIPS64 has not seen much use lately.

**Comment by martin-frbg** [comment]
Copying kernel/x86/KERNEL.generic to kernel/mips64 should solve this. (At least it does on the small Ubiquiti router I use for testing) BTW we seem to have conflicting reports about whether current OpenBLAS builds and works on Loongson3A - #1776 vs #902

**Comment by XiWeiGu** [comment]
Thanks, i disable the assembly optimization on loongson 3A4000 successfully.
 And when i try to enable msa optimizatios on loongson 3A4000 with gcc7.3.1, i got the following error message:
`../kernel/mips/../mips/daxpy_msa.c: In function ‘daxpy_k’:
../kernel/mips/../mips/dscal_msa.c:47:28: warning: implicit declaration of function ‘__msa_cast_to_vector_double’ [-Wimplicit-function-declaration]
             v2f64 zero_v = __msa_cast_to_vector_double(0);
../kernel/mips/../mips/macros_msa.h:74:11: warning: implicit declaration of function ‘__msa_cast_to_vector_double’ [-Wimplicit-function-declaration]
     out = __msa_cast_to_vector_double(a);          \
../kernel/mips/../mips/daxpy_msa.c:75:18: note: in expansion of macro ‘COPY_DOUBLE_TO_VECTOR’
         da_vec = COPY_DOUBLE_TO_VECTOR(da);
../kernel/mips/../mips/macros_msa.h:74:9: error: incompatible types when assigning to type ‘v2f64 {aka __vector(2) double}’ from type ‘int’
     out = __msa_cast_to_vector_double(a);          \
../kernel/mips/../mips/daxpy_msa.c:75:18: note: in expansion of macro ‘COPY_DOUBLE_TO_VECTOR’
         da_vec = COPY_DOUBLE_TO_VECTOR(da);
../kernel/mips/../mips/dscal_msa.c:47:28: error: incompatible types when initializing type ‘v2f64 {aka __vector(2) double}’ using type ‘int’
In file included from ../kernel/mips/../mips/dscal_msa.c:29:0:
../kernel/mips/../mips/macros_msa.h:74:9: error: incompatible types when assigning to type ‘v2f64 {aka __vector(2) double}’ from type ‘int’
     out = __msa_cast_to_vector_double(a);          \`
i can not find the `__msa_cast_to_vector_double`  in msa 1.0 and msa 2.0.

**Comment by XiWeiGu** [comment]
Is every pull request checked on the mips /mips64 platform?

**Comment by martin-frbg** [comment]
There is no CI setup for mips, I only do test builds on two small routers occasionally. There has not been much work on any of the MIPS kernels in recent years anyway (and I bought the cheap routers specifically to make sure that the MIPS codes still build at all). From what I could find, the __msa_cast_to_vector_double seems to be a gcc shortcut for __builtin_msa_cast_to_vector_double and should be in the msa.h included from macros_msa.h

**Comment by XiWeiGu** [comment]
i can not find  the `__builtin_msa_cast_to_vector_double` in [https://gcc.gnu.org/onlinedocs/gcc/MIPS-SIMD-Architecture-Built-in-Functions.html#MIPS-SIMD-Architecture-Built-in-Functions](url).

**Comment by XiWeiGu** [comment]
also, i can not find in msa.h.

**Comment by martin-frbg** [comment]
The LLVM msa.h has it, and I find references to it in the android sdk for mips64 and in what looks like an old (2014?) gcc tree in the chromiumos sources online. Perhaps dropped or moved to another include at some point in recent gcc development history ?

**Comment by martin-frbg** [comment]
Hmm, weird. Apparently the MSA support was merged into gcc only in May 2016 (and the msa.h in that commit did not have the cast_to_vector... macros either). Perhaps the "other" version came from some vendor branch that was never merged into mainline gcc ? (All google references point into gcc "4.9" versions included in either chromium or android trees...)

**Comment by martin-frbg** [comment]
Also, is the Loongson3A4000 fundamentally different from the MIPS64 Loongson3A target that OpenBLAS already has ? (Given that xianyi coded it in 2011, that would have been something like
the Loongson3A1000 if wikipedia is to be trusted - perhaps the target should be renamed to clarify
what hardware it supports ?)

**Comment by XiWeiGu** [comment]
It's not completely different, the following is the cpuinfo in Loongson3A4000:
system type		: generic-loongson-machine
machine			: loongson,generic
processor		: 0
cpu model		: ICT Loongson-3 V0.0  FPU V0.1
model name		: Loongson-3A R4 (Loongson-3A4000) @ 1800MHz
BogoMIPS		: 3594.02
wait instruction	: yes
microsecond timers	: yes
tlb_entries		: 2112
extra interrupt vector	: no
hardware watchpoint	: yes, count: 0, address/irw mask: []
isa			: mips1 mips2 mips3 mips4 mips5 mips32r1 mips32r2 mips64r1 mips64r2
ASEs implemented	: vz msa
shadow register sets	: 1
kscratch registers	: 6
package			: 0
core			: 0
VCED exceptions		: not available
VCEI exceptions		: not available

this is the Makefile.conf on Loongson3A4000:
OSNAME=Linux
ARCH=mips64
C_COMPILER=GCC
BINARY32=
BINARY64=1
CEXTRALIB=-L/opt/rh/devtoolset-7/root/usr/lib/gcc/mips64el-redhat-linux/7 -L/opt/rh/devtoolset-7/root/usr/lib/gcc/mips64el-redhat-linux/7/../../../../lib64 -L/lib64/../lib64 -L/usr/lib64/../lib64 -L/opt/rh/devtoolset-7/root/usr/lib/gcc/mips64el-redhat-linux/7/../../.. -L/lib64 -L/usr/lib64  -lc
F_COMPILER=GFORTRAN
FC=gfortran
BU=_
NOFORTRAN=1
CORE=LOONGSON3A
LIBCORE=loongson3a

**Comment by martin-frbg** [comment]
Thanks. I was mainly asking due to the confusion in #1776, but now that I reread it the original problem there must have been related to cross-compiling on some unspecified other hardware. (So probably some mixup of host and target tools occured - using HOSTCC=gcc and CC=gcc without a path to the cross-compiler does not look good).

**Comment by 212dandan** [comment]
Hello, I am an employee of LoongsonTechnology Co, Ltd. 
We intend to give you a Loongson3A4000 machine to test and solve the problems on MIPS kernel, and solve gcc compilation problems. 
Could you give me a valid email address to sign the agreement, and also provide the recipient address, phone number, etc. We can talk about the specific address and other details in the email.
Finally, my email address is zhangdandan@loongson.cn.
Thanks.

**Comment by xianyi** [comment]
Hi @212dandan , @gxw-loongson 

I developed the kernel for Loongson 3A a few years ago. And I have compiled OpenBLAS on 3A3000 machine successfully. Please try OpenBLAS loongson3a branch.

**Comment by xianyi** [comment]
Hi @212dandan , @gxw-loongson 

I developed the kernel for Loongson 3A a few years ago. And I have compiled OpenBLAS on 3A3000 machine successfully. Please try OpenBLAS loongson3a branch.

**Comment by martin-frbg** [comment]
@xianyi I think their main problem now is that the MSA parts may have relied on features from some vendor branch that never made it into the mainline gcc. 
I have signed the agreement in case you would not have time, but the hardware has not shipped yet. If you would like to take it, it would certainly simplify logistics a lot.

**Comment by martin-frbg** [comment]
Closing as fixed by gxw-loongson in #3027
